### PR TITLE
Implement support for hostpath solr data volumes.

### PR DESCRIPTION
**Pull request by jward-bw**
This PR adds a new option for ephemeral data storage. Instead of the emptyDir you can use a `hostPath` volume. The default behaviour for ephemeral storage is still emptyDir, but specifying `HostPath` will mean hostPath will be used instead of emptyDir.

Addresses https://github.com/apache/solr-operator/issues/266

### CODE REVIEW

**Comment by jward-bw** [comment]
@HoustonPutman I'm not sure what this error means, any ideas?
https://github.com/apache/solr-operator/pull/265/checks?check_run_id=2440577547

**Comment by HoustonPutman** [comment]
> @HoustonPutman I'm not sure what this error means, any ideas?
> https://github.com/apache/solr-operator/pull/265/checks?check_run_id=2440577547

That is saying that you have two extra lines in your go.mod.

Try running `make mod-tidy` and see if those go away.

**Comment by jward-bw** [comment]
They don't go away. Should I just remove the changes to go.mod?

**Comment by HoustonPutman** [comment]
That's a good idea. I pulled your PR and ran `make mod-tidy`, and those two lines were removed. Maybe just a local thing for you?

**Comment by jward-bw** [comment]
Interestingly I removed both and re-ran make mod-tidy,  and the `github.com/elazarl/goproxy/ext` one came back, but the other didn't. I've removed both anyway

**Comment by jward-bw** [comment]
I don't use go regularly, so I shouldn't have any weird config. I did recently update to the latest version, so I wonder if that's caused a change in behaviour fo rme?

File: controllers/util/solr_util.go [hunk 1]
```diff
@@ -214,16 +214,20 @@ func GenerateStatefulSet(solrCloud *solr.SolrCloud, solrCloudStatus *solr.SolrCl
...
-		emptyDirVolume := corev1.Volume{
-			Name: solrDataVolumeName,
-			VolumeSource: corev1.VolumeSource{
-				EmptyDir: &corev1.EmptyDirVolumeSource{},
-			},
+		ephemeralVolume := corev1.Volume{
+			Name:         solrDataVolumeName,
+			VolumeSource: corev1.VolumeSource{},
 		}
 		if solrCloud.Spec.StorageOptions.EphemeralStorage != nil {
```
Comment (HoustonPutman) [inline controllers/util/solr_util.go:221]: If `solrCloud.Spec.StorageOptions.EphemeralStorage` is empty, but not `nil`, then I think we will have an empty VolumeSource. We want to default to a plain EmptyDir in this case as well.

**Review by HoustonPutman — CHANGES_REQUESTED** [review]
This looks very good. One small issue with an empty ephemeral storage options. Not sure if it could actually happen, depending on how go parses the yaml/json, but still good to protect against anyways.

Also would you mind adding this option to the "Data Storage" section in `docs/solr-cloud/solr-cloud-crd.md`?

File: controllers/util/solr_util.go (hunk 1 above)
Comment (jward-bw) [reply controllers/util/solr_util.go:221]: I might have misunderstood what you meant, but I've added a check for if EmptyDir is nil, and a corresponding test. There was already a test for if EphemeralStorage is empty though (if that's what you meant?):
https://github.com/apache/solr-operator/blob/c78b2b80ab097cb7f1b573f4fca8903ba7f63eb9/controllers/solrcloud_controller_storage_test.go#L240

Comment (HoustonPutman) [reply controllers/util/solr_util.go:221]: Ahh wonderful. Must have missed the test. But good change on the code 👍

File: controllers/solrcloud_controller_storage_test.go [hunk 2]
```diff
@@ -329,6 +449,58 @@ func TestEphemeralStorageWithSpecs(t *testing.T) {
...
+		mgrStopped.Wait()
+	}()
+
+	cleanupTest(g, instance.Namespace)
+
+	// Create the SolrCloud object and expect the Reconcile and StatefulSet to be created
+	err = testClient.Create(context.TODO(), instance)
+	g.Expect(err).NotTo(gomega.HaveOccurred())
+	defer testClient.Delete(context.TODO(), instance)
+	g.Eventually(requests, timeout).Should(gomega.Receive(gomega.Equal(expectedCloudRequest)))
```
[the remark this answers is not in the document]
Comment (HoustonPutman) [reply controllers/solrcloud_controller_storage_test.go:485]: (Also no need to force push each time, I'll squash-merge in the end.)

[the remark this answers is not in the document]
Comment (jward-bw) [reply controllers/solrcloud_controller_storage_test.go:485]: Ok thanks, I've been on repos before where they like to have everything in a single commit :)

**Review by HoustonPutman — CHANGES_REQUESTED** [review]
The looks good to me!

One final request:
Would you mind adding this option to the "Data Storage" section in `docs/solr-cloud/solr-cloud-crd.md`?

**Comment by jward-bw** [comment]
> The looks good to me!
> 
> One final request:
> Would you mind adding this option to the "Data Storage" section in `docs/solr-cloud/solr-cloud-crd.md`?

Done! Sorry I missed that yesterday.

**Review by HoustonPutman — APPROVED** [review]
Looks good! Thanks for the great contribution!

**Pull request merged**